### PR TITLE
Convert attacker retreat steps

### PR DESCRIPTION
**Pull request by trevan**
This converts the attacker retreat steps into a single battle step.

## Testing

Tested partial amphibious retreats and regular retreats.  Loaded an old save that was in the middle of a partial amphibious retreat.

### CODE REVIEW

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java [hunk 1]
```diff
@@ -1661,45 +1667,29 @@ public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
...
-                    attackingUnits, gameData, isAmphibious)) {
-              attackerRetreatNonAmphibUnits(bridge);
-            }
-          }
-        });
-    steps.add(
-        new IExecutable() {
-          private static final long serialVersionUID = 669349383898975048L;
+      @Override
+      public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {}
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:1676]: Codacy found an issue: [Document empty method body](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6178813)

**Comment by panther2** [comment]
@trevan 
Does that address #4688 and #4821, too?

**Comment by trevan** [comment]
> @trevan
> Does that address #4688 and #4821, too?

@panther2 , no, this doesn't change any functionality of the game.  It is just refactoring the code.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java [hunk 2]
```diff
@@ -1661,45 +1667,33 @@ public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
...
-              attackerRetreatNonAmphibUnits(bridge);
-            }
-          }
-        });
-    steps.add(
-        new IExecutable() {
-          private static final long serialVersionUID = 669349383898975048L;
+      @Override
+      public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+        // Intentionally left blank
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:1677]: Some background on why this is intentional could be useful. Why is this intentionally left blank?

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  @Override
+  public Order getOrder() {
+    return Order.OFFENSIVE_GENERAL_RETREAT;
+  }
+
+  @Override
+  public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+    retreatUnits(bridge);
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:50]: It's really odd for `retreatUnits` to be the implementation of this interface method, and also called externally directly. Can you speak to why that is @trevan ?

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Value;
+import org.triplea.java.collections.CollectionUtils;
+
+@AllArgsConstructor
+public class OffensiveGeneralRetreat implements BattleStep {
+
+  final BattleState battleState;
+
+  final BattleActions battleActions;
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:30]: Can these variables be made private?

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+  @Override
+  public List<String> getNames() {
+    if (canAttackerRetreat()
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:34]: Nitpicky - Even if a multi-line, I still think this would be cleaner as a ternary. I think with context, it seems easier, but I would suggest that maintaining/reading the code, it's easier to grok as a ternary. Perhaps the boolean can be extracted to a helper method to make it a simple tenary. WDYT?

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    return Order.OFFENSIVE_GENERAL_RETREAT;
+  }
+
+  @Override
+  public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+    retreatUnits(bridge);
+  }
+
+  public void retreatUnits(final IDelegateBridge bridge) {
+    if (battleState.isOver()) {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:54]: Does this condition happen often? TBH, it seems a bit like a bug. If the battle is over, the caller perhaps did not realize that and should not have made this call to begin with.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        battleState.isAmphibious());
+  }
+
+  private boolean canAttackerRetreatSeaPlanes() {
+    return battleState.getBattleSite().isWater()
+        && battleState.getUnits(BattleState.Side.OFFENSE).stream().anyMatch(Matches.unitIsAir());
+  }
+
+  private boolean canAttackerRetreatPartialAmphib() {
+    if (Properties.getPartialAmphibiousRetreat(battleState.getGameData())) {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:108]: Have you consider pushing this conditional to the callers? I think in both cases it would be the calling code easier to understand, it avoids having to jump around the class file to understand those locations. Furthermore, without this if statement the method body would be a simple stream & any match statement. Not that it matters too much, there would be a small performance boost as you wouldn't need to compute the full set of land units but could halt on the first match.

**Comment by DanVanAtta** [comment]
Did a first pass, looks pretty solid @trevan ; Will try to come back to this tomorrow to finish the review.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java (hunk 2 above)
Comment (trevan) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:1677]: For some reason, the retreat steps were broken up into three separate IExecutables even though only one of them would ever run.  I merged all of them into one step and so two of the old IExecutables no longer have anything to do.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java (hunk 3 above)
Comment (trevan) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:50]: The same logic (retreat) is called in other steps.  I either had the choice of creating a static helper class to store the logic or to allow the other steps to instantiate and call this method directly.

I would prefer that the other steps didn't need this logic and hope that I might be able to fix it but that will involve changing the logic of the steps and might even change the game play.  I'd like to make that change in a separate PR.

Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:50]: Fair enough. Extracting logic for re-use would seemingly be okay. Passing an instance to a constructor can help avoid static coupling.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java (hunk 6 above)
Comment (trevan) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:54]: Yes.  How the stack work is that at the beginning of the round, all of the IExecutables/BattleSteps are created and added to the stack.  Then, it goes through each of the items in the step.  There isn't currently a way to short circuit the stack when the battle is over.  So, this step already is in the stack and will execute even if the battle finished in a previous step.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java (hunk 7 above)
Comment (trevan) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:108]: I think it is better if the caller doesn't have to know what properties enable/disable partial amphibious retreat.

I've made the stream/anyMatch change, though.

Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java]: Hmm, it's a private method though.

Generally it's better to get the conditional logic done at a higher layer. So you have something that looks like this:

```
public void method() {
  if(a || b) {
        doFoo();
         if(a1) {
               doBar()
          }
  } else if(c) {
      doZed();
  }
}
```

vs:

```
public void method() {
   doFoo();
   doBar();
   doZed();
}

private void doFoo() {    
    if(!a && !b && c) {
        doZed();
    } else if(!a && !b) {
        return;
    }
    doBar();
}

private void doBar() {
    if(!a1) {
        return;
    }
}
```

Notice that above the control flow is pushed down and it's very hard to follow what the logic is.

The idea is 3 part:
1. Avoid lists of void method calls
2. Choose where the control flow will be, have the method with control flow just do mainly that and the things it invokes should be relatively functional.
3. Push control flow to higher layers, avoid doing control flow in private methods That indicates you perhaps want a strategy object to handle the behavior.

Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java]: I'm not insisting any changes be made here, but I don't agree that having control flow at a lower layer like this makes it easier. A reader of the code needs to jump to this method anyways and it's cognitively complex to keep track of control flow throughout  'n' layers of code.

Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java]: Expouding on "needs to jump to this method anyways"

Let's say the reader knows that only "a" is true. In this case the following method:

```
public void method() {
  if(a || b) {
        doFoo();
         if(a1) {
               doBar()
          }
  } else if(c) {
      doZed();
  }
}
```

Requires a reader to only read `doFoo`. On the other hand, with the series of void call, a reader needs to go through each method, hence why I suggest often they may as well be inlined, as you have to read it all anyways. The fact you can't skip code (as in this example where you can skip 'doBar' and 'doZed'), is a break-down of information hiding. My 2 cents.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java (hunk 6 above)
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:55]: Thanks for the context. I don't find that terribly satisfying, but I would accept that wiring a 'Predicate' for battle is over to the execution stack at this juncture is potentially not easy and out of scope of this update.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java (hunk 2 above)
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:1677]: Can you update the code with a comment or a 'remove on major release' annotation? This would help us to know what to do with this going forward.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java [hunk 8]
```diff
@@ -20,24 +19,26 @@
...
-  final BattleState battleState;
+  private final BattleState battleState;
 
-  final BattleActions battleActions;
+  private final BattleActions battleActions;
 
   @Override
   public List<String> getNames() {
-    if (canAttackerRetreat()
+    return isRetreatPossible()
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:32]: :+1: 

This looks good, thanks!

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java [hunk 9]
```diff
@@ -105,18 +106,13 @@ private boolean canAttackerRetreatSeaPlanes() {
...
-        }
-      }
+    if (!Properties.getPartialAmphibiousRetreat(battleState.getGameData())) {
+      return false;
     }
-    return false;
+    // Only include land units when checking for allow amphibious retreat
+    return battleState.getUnits(BattleState.Side.OFFENSE).stream()
+        .filter(Matches.unitIsLand())
+        .anyMatch(unit -> !unit.getWasAmphibious());
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java:115]: nitpicky, note - you can use Predicates here to write this as a method references:
```
.anyMatch(Predicates.not(Unit::getWasAmphibious));
```

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java (hunk 7 above)
Comment (trevan) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java]: I think I understand what you are saying but this entire method is control flow.  So it is like:
```
public void method() {
  if (isAllowed()) {
    doFoo();
  }
}

private boolean isAllowed() {
  return a || (b && c) || d;
}
```
vs
```
public void method() {
  if (a || (b && c) || d) {
    doFoo();
  }
}
```

You just recommended that I do the former in `getNames()` (with the the new `isRetreatPossible()` method).  How is `canAttackerRetreatPartialAmphib()` any different?

Or is it because both `canAttackerRetreatPartialAmphib()` and `canAttackerRetreatAmphibPlanes()` checks for `Properties.getPartialAmphibiousRetreat` and so you want to move that call out of them?

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java (hunk 2 above)
Comment (trevan) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:1700]: I added the annotation to all of the `execute` methods of the old `IExecutables`.  I also add a larger comment to indicate why it is blank.

**Comment by trevan** [comment]
The build failure is unrelated to my changes: `Exception in thread "main" javax.net.ssl.SSLProtocolException: Connection reset`

**Comment by DanVanAtta** [comment]
marking for further review later, in the meantime I'm pretty sure there is nothing here that we can't simply follow up on later.
